src/Dfc.CourseDirectory.WebV2/Features/Venues/DeleteVenue/DeleteVenue.cs

```diff
@@ -0,0 +1,232 @@
+
+            await Task.WhenAll(getCourses, getApprenticeships, getTLevels);
+
+            return CreateViewModel(venue, request.ProviderId, await getCourses, (await getApprenticeships).Values, await getTLevels);
+        }
+
+        public async Task<OneOf<NotFound, ModelWithErrors<ViewModel>, Success>> Handle(Command request, CancellationToken cancellationToken)
+        {
+            var queryResult = await Handle(new Query { VenueId = request.VenueId, ProviderId = request.ProviderId }, cancellationToken);
+
+            return await queryResult.Match(
+                notFound => Task.FromResult<OneOf<NotFound, ModelWithErrors<ViewModel>, Success>>(notFound),
+                async vm =>
+                {
+                    var validationResult = await new CommandValidator(vm).ValidateAsync(request);
+
+                    if (!validationResult.IsValid)
+                    {
+                        return new ModelWithErrors<ViewModel>(vm, validationResult);
+                    }
+
+                    var result = await _cosmosDbQueryDispatcher.ExecuteQuery(new Core.DataStore.CosmosDb.Queries.DeleteVenue { VenueId = request.VenueId });
+
+                    if (!(result.Value is Success))
+                    {
+                        return new NotFound();
+                    }
+
+                    _providerOwnershipCache.OnVenueDeleted(request.VenueId);
+
+                    _journeyInstanceProvider.GetOrCreateInstance(() => new JourneyModel { VenueName = vm.VenueName })
+                        .Complete();
+
+                    return new Success();
+                });
+        }
+
+        public Task<DeletedViewModel> Handle(DeletedQuery request, CancellationToken cancellationToken)
+        {
+            var journeyInstance = _journeyInstanceProvider.GetInstance<JourneyModel>();
+            journeyInstance.ThrowIfNotCompleted();
+
+            return Task.FromResult(new DeletedViewModel
+            {
+                VenueName = journeyInstance.State.VenueName
+            });
+        }
+
+        private static ViewModel CreateViewModel(Venue venue, Guid providerId, IEnumerable<Course> courseRuns, IEnumerable<Apprenticeship> apprenticeships, IEnumerable<SqlModels.TLevel> tLevels) => new ViewModel
+        {
+            VenueId = venue.Id,
+            ProviderId = providerId,
+            AffectedCourses = GetAffectedCourses(courseRuns, venue.Id),
+            AffectedApprenticeships = GetAffectedApprenticeships(apprenticeships, venue.Id),
+            AffectedTLevels = GetAffectedTLevels(tLevels, venue.Id),
+            VenueName = venue.VenueName,
+            AddressParts = new[]
+            {
+                venue.AddressLine1,
+                venue.AddressLine2,
+                venue.Town,
+                venue.County
+            }.Where(part => !string.IsNullOrWhiteSpace(part)).ToArray(),
```

Comment:
repeats logic from elsewhere, stronger argument for pulling out and unit testing, see my other comment

Reply:
Same argument from earlier. I diagree - it's ViewModel related logic and we shouldn't be sullying a domain/storage model with it. I'd also advocate maintaining a separation of concerns in View logic across features in this instance as these may diverge in future.

Reply:
That's an interesting thought about divergence. I suspect the translation for display would be expected to be consistent across the whole UI, so maybe something more explicitly shared would make sense. Perhaps a question for Westley about the intent.